### PR TITLE
Ocm

### PR DESCRIPTION
**Pull request by TheDiemer**
Adding in the capability to call OCM and convert a cluster ID into a SFDC account ID using said backend.

### CODE REVIEW

File: proactive-support-bot.py [hunk 1]
```diff
@@ -56,6 +57,13 @@ def handle_fileuploadtest(client, event):
         channels=channel,
         content="Hello, World")
 
+def handle_cluster(client, event):
+    channel = event['event']['channel']
+    cluster = event['event'].get('text').split()[-1]
+    account = ocm.cluster_to_account(os.environ['PYOCM_SECRETpyocm'], cluster)
```
Comment (wking) [inline proactive-support-bot.py]: `PYOCM_SECRETpyocm` -> `PYOCM_SECRET`?

Comment (wking) [reply proactive-support-bot.py]: Or `PYOCM_TOKEN`?

Comment (TheDiemer) [reply proactive-support-bot.py]: It is the way the secret got added as the environment variable. I'm happy to modify it. hang tight and I'll go update it for the pod

Comment (wking) [reply proactive-support-bot.py]: and maybe pull it out as a module global:

```python
ocm_token = os.environ['OCM_TOKEN']
```
and then here:
```python
account = ocm.cluster_account(token=ocm_token, cluster=cluster)
```

File: proactive-support-bot.py [hunk 2]
```diff
@@ -56,6 +57,13 @@ def handle_fileuploadtest(client, event):
...
+    return client.chat_postMessage(
+            channel=channel,
+            text='Based on {0}, we found SFDC Account {1}'.format(cluster, account))
```
Comment (wking) [inline proactive-support-bot.py]: What is SFDC?  Can we spell it out?

File: ocm.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
+#!/bin/python
+import requests
+
+def call_ocm(endpoint, token, parameters=None):
+    requests.packages.urllib3.disable_warnings()
```
Comment (wking) [inline ocm.py]: this should happen at the module scope, before this function definition.  We don't need to call it in every `call_ocm`, do we?

File: ocm.py [hunk 4]
```diff
@@ -0,0 +1,44 @@
+#!/bin/python
```
Comment (wking) [inline ocm.py]: You don't need the shebang for this module, it is never called directly.

File: ocm.py [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+import requests
+
+def call_ocm(endpoint, token, parameters=None):
+    requests.packages.urllib3.disable_warnings()
+    r = requests.get("{}/{}".format("https://api.openshift.com/api/", endpoint),
+            params=parameters, 
+            headers = {
+                'Accept': 'application/json',
+                'Authorization': 'Bearer {}'.format(token),
+                'User-Agent': 'pythonCall'
```
Comment (wking) [inline ocm.py:11]: let's be nice and add [a version][1].  Module-level global:

```python
__version__ = '0.1.0'
```
and here:
```python
'User-Agent': 'cluster-support-bot/{}'.format(__version__),
```
or some such?

[1]: https://tools.ietf.org/html/rfc2616#section-3.8

File: ocm.py [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+                }
+            )
+
+    if r.status_code == 204:
+        return []
```
Comment (wking) [inline ocm.py:16]: Can we just `return` (with the implicit `None`) here?  A mutable list seems potentially surprising :p

File: ocm.py [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+    if r.status_code != 200:
+        raise Exception('''looking up infomation from: {}\n
+                Error Code: {}\n {}'''.format(endpoint, r.status_code, r.url))
```
Comment (wking) [inline ocm.py:19]: can we raise `ValueError(r)` here?  Or some more-specific error?  And then catch that and log it in `handle_cluster`?

File: proactive-support-bot.py (hunk 2 above)
Comment (TheDiemer) [reply proactive-support-bot.py]: Salesforce, I expanded it to Red Hat Customer Account as that is technically more accurate.

File: proactive-support-bot.py (hunk 1 above)
Comment (TheDiemer) [reply proactive-support-bot.py]: Both are great points, handled in recent commit.

File: ocm.py (hunk 6 above)
Comment (TheDiemer) [reply ocm.py:16]: Sounds like a great idea to me!

File: ocm.py (hunk 5 above)
Comment (TheDiemer) [reply ocm.py:11]: Following RFC, sounds like a good idea :+1:

File: ocm.py (hunk 7 above)
Comment (TheDiemer) [reply ocm.py:19]: Sorry, not super familiar with what you are asking for here. Do you mind elaborating a bit @wking ?

File: ocm.py [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+
+    return r.json()
+
+
+def cluster_to_account(token, cluster):
+    requests.packages.urllib3.disable_warnings()
```
Comment (wking) [inline ocm.py:25]: You can drop this now that you call it at the module level

Comment (TheDiemer) [reply ocm.py:25]: D: I thought I pulled that one! crap.

File: ocm.py (hunk 7 above)
Comment (wking) [reply ocm.py:19]: this code is down in a low-level, and it's doing some string formatting to set things up for eventually being logged further up (or just crashing the script ;).  I'd rather skip the string formatting and leave the data structured down here:

```python
if r.status_code != 200:
    raise ValueError({'response': r})
```

and then up in `handle_cluster`:

```python
try:
    account = ocm.cluster_to_account(token=ocm_token, cluster=cluster)
except ValueError as error:
    if len(error.args) == 1:
        details = error.args[0]
    else:
        logger.error('unrecognized cluster_to_account error: {}'.format(error))
        return

    response = details.get('response')
    if not response:
        logger.error('cluster_to_account error had no request: {}'.format(error))
    return client.chat_postMessage(
            channel=channel,
            text='Failed to find {}: {} -> {} {}'.format(cluster, response.url, response.status_code, response.body.read())
```
or some such.

**Comment by wking** [comment]
I've pushed https://github.com/wking/proactive-support-bot/tree/ocm-v2 ( 9aa2e5e685333 ) with a few touch-ups here to pull in my remaining suggestions.

**Comment by TheDiemer** [comment]
I am willing to kill this PR in favor of your code @wking That way we don't need to pull in my code separately causing a need for a rebase.

Instead I will begin modelling the hydra integration on your ocm code as well. Thanks for all the pointers!

**Pull request closed without merging**